### ISSUE TITLE
Should same-key-twice bindings be restored?

### ISSUE THREAD

**Issue by tarsius** [magit/magit#2435]
> https://www.reddit.com/r/emacs/comments/3w2yo8/magit_key_commands_changed/

**Comment by proofit404** [comment]
+1 for Dwim

**Comment by jwiegley** [comment]
Hello to everyone. First, thanks to @tarsius for bringing up this UI concern in the first place.

My suggestion thus far is to make "safe mode" the default, so that those new to Magit will not shoot themselves in the foot.  But perhaps we can also have an option for "expert mode", where shotguns are liberally handed out for extreme foot removal.  This would restore `P P` and `F F`, for example, but with no guarantees made as to correctness of the underlying operation.

Meanwhile, safe mode defaults help make users more conscious of the possibilities, and why they should think before taking one action over another.

**Comment by manuel-uberti** [comment]
I'd go with safe. As much as `P P` and `F F` never went wrong for me, better safe than sorry.

**Comment by Silex** [comment]
Ah, just noticed the new popups for pull, push & branches. My initial reaction was like "nooooo I'm used to them" but then I saw that these popups seems better thought out overall and are much clearer... the cyclable options are also nice.

**Comment by lunaryorn** [comment]
I'm on the safe side, too, and I really like these new popups.  I find that they much more resemble my personal mental model of how Git works.

**Comment by philjackson** [comment]
My 0.2c is; I really like the new layout - makes more sense logically, but, I think the DWIM stuff should have stayed and be enabled by default. I don't think "it's locked into my muscle memory" is a necessarily a poor argument for keeping bindings as they were - we're Emacs users, after all!

As a side-note, there was a lot of fuss when I changed push from `P` to `P P`, I was running out of keys though...

Keep up the amazing work, don't let these moments of angry opinions demotivate you.

**Comment by tarsius** [comment]
```lisp
(defun magit-dwim-oldschool ()
  "Dwim like in the olden days.

Modify bindings in the fetch, pull, rebase, and push popups.
Restores the same-key-twice bindings which were traditionally
used for the most commonly used variant available in a popup."
  (magit-change-popup-key 'magit-fetch-popup  :action ?u ?f)
  (magit-change-popup-key 'magit-pull-popup   :action ?u ?F)
  (magit-change-popup-key 'magit-rebase-popup :action ?e ?r)
  ;; Using "PP" for "push to upstream" no longer makes sense,
  ;; because doing that would be a mistake most of the time.
  (magit-change-popup-key 'magit-push-popup   :action ?p ?P)

  ;; MAYBE
  ;; "re" is available again, restore old binding for consistency
  ;; with "ce" and at the cost of breaking consistency with "[fFP]e".
  ;; (magit-change-popup-key 'magit-push-popup   :action ?m ?e)
  ;; Since we are restoring old bindings, we might as well restore
  ;; this one too.
  ;; (magit-change-popup-key 'magit-push-popup   :action ?s ?o)
  )
```

**Comment by tarsius** [comment]
I'll respond tomorrow. To tired now - boring football match - must keep eyes open.

**Comment by tarsius** [comment]
> My suggestion thus far is to make "safe mode" the default, so that those new to Magit will not shoot themselves in the foot. But perhaps we can also have an option for "expert mode", where shotguns are liberally handed out for extreme foot removal.

Actually I don't think it's new users who are most in danger, but those who have been using Magit for a long time and hardly ever fall back to the command line anymore. In some cases what Magit has to offer isn't so great and when I improve these parts I sometimes get the impression I have to deal with a bit of Stockholm syndrome. ;-) To me pre-v2.1 pushing was such an area. I always did that on the command line because Magit's implementation seemed so inadequate. That changed when I implemented the v2.1 variant (which happened way before v2.1 was released, but still long after had become the maintainer). It was always a mystery to me why no one complained about the pushing. 

I also don't think "expert mode" is really the right word when talking about these key bindings. An expert and a beginner alike might prefer either the same-key-twice bindings she is used to, or the new bindings which use the same keys in all popups that offer "to/from upstream/push-remote/elsewhere" variants.

(By the way I believe to remember that you were among those argued *against* using <kbd>r r</kbd> to *continue* a rebase! And I argued that same-key-twice bindings should be used for the default action whenever possible :-)

----

But aside from terminology I agree. And I expected I would have to provide some way of getting the traditional bindings back, but didn't want to do it preemptively. I was hoping some good way of doing that would be suggested while the feature has not been merged into `master` yet. Unfortunately I got very little feedback (except code review) before merging to `master`. So I had to merge in order to get the needed feedback.

**Comment by tarsius** [comment]
```lisp
(magit-change-popup-key 'magit-fetch-popup  :action ?u ?f)
(magit-change-popup-key 'magit-pull-popup   :action ?u ?F)
(magit-change-popup-key 'magit-rebase-popup :action ?e ?r)
(magit-change-popup-key 'magit-push-popup   :action ?p ?P)
```

Is the minimal, very `define-key`-like, way of getting most of the "dwim" bindings back. That should be made available in a more convenient way though, maybe using a function as above or more likely using an option. (It also has be more clear on how <kbd>P P</kbd> now means something different from before.) Or does anyone have a better suggestion?

Full disclosure: I also changed the bindings of the "create and/or checkout branch" commands. In that case the change was completely "unnecessary". If you don't like that change, then you should say so now.

**Comment by tarsius** [comment]
I also don't think "dwim" is the right term. <kbd>P P</kbd> did never behave differently (but "as one would expect") depending on context. It used to push to the branch that was configured as the upstream, and now that Magit supports a separate push-remote, it should push to that. But this binding did never guess what you meant, instead it did what you previously configured it to do. (Which didn't have to be the same thing for different branches, but *configurable* and *dwim* are not the same concept.)

<kbd>P P</kbd> never dwimmed like e.g. <kbd>d d</kbd> does. The latter shows the diff for "the thing at point", and therefore is a true dwim command. Doing the same for <kbd>P P</kbd> would mean that it pushes to `origin/master` when point is on the *Unpushed to origin/master* section, and would push to `my-fork/new-feature` when point is on the *Unpushed to my-fork/new-feature*.

That sounds very dangerous. For diffing we can take that risk because if the wrong changes changes are shown, then that causes no harm. But if you push some crap to `master`, then you might be spanked, depending on the policy of your project.

The point I am trying to make is this: call the thing you want to be restored by its name. *What you want is to be able to press the same key twice to perform the most commonly used variant.* That's a legitimate preference and there is no need to label it as "expert mode" and "dwim functionality".

Going from there, you should consider whether you want all of these bindings <kbd>F F</kbd>, <kbd>f f</kbd>, <kbd>r r</kbd>, and <kbd>P P</kbd> restored or only <kbd>F F</kbd> and/or <kbd>P P</kbd>. 

Also take into account that different users will likely have different opinions on that, which means I probably cannot provide a "go back to how it used to be" function/option. For that reason I am currently considering to *instead* write a tutorial "Branching in Magit", which would, toward the end, mention what key bindings were changed and provide the *code* necessary to do so.

I am most sympathetic to the desire to get <kbd>P P</kbd> back. Unfortunately this is also the binding which is most problematic because, if restored, it would do something *different* from what it used to do. It used to push to the upstream, but now it should push to the push-remote.

That's why a tutorial is needed. Most Git and Magit users, even though they do use a triangular work-flow in practice, are not aware that Git supports that by making it possible to configure *two* related branches. Even though the English word "upstream" quite strongly implied what the "upstream branch" is, many users abused it as the place they push to.

That is understandable, they did not know that Git also supports a separate push-remote and because they found it more useful to configure the branch they push to as "*the* related branch" that's what they did. But now that Magit does support the push-remote, this has to be unlearned.
<kbd>P P</kbd> should push to the configured push-remote, because once that is configured that is where one wants to push to in 99% of all cases.

Some of you might consider that <kbd>P P</kbd> should do that *provided* the push-remote is configure, and else it should *fall back* to push to the upstream. That would be *very* dangerous, because even if you do understand the push-remote/upstream distinction and fully agree, that one should push to the push-remote, one might still have some old repositories laying around, in which that has not yet been configured.

I'll talk about the other lost same-key-twice bindings some other time.

**Comment by jwiegley** [comment]
@tarsius You make some good points. Go with what you believe to be right, and I'll locally bind my keys to suit my tastes.

**Comment by pronobis** [comment]
Good points indeed, I just would like to add that since the popup does not list the `p` command if push-remote is not configured, it's only after reading this that I noticed that `P p` is an option. Would be even cooler if it was possible to configure the push-remote for this branch the first time this option is used (if it was not configured before)

**Comment by tarsius** [comment]
I have remove `branch.autoSetupPush` again, see 1890dfa808ec1b38a035eac9098e2d8a55d0da53.

**Comment by tarsius** [comment]
@proofit404 That might work. What do others think?

If both the upstream and the push-remote are set, then display:

```
Push feature to
 p tarsius/feature
 u origin/master
 e elsewhere
```

When they are both missing, then display:

```
Push feature to
 p new @{push}
 u new @{upstream}
 e elsewhere
```

(I am struggling finding a short phrase that makes sense following "Push feature to". What I have used above is grammatically correct, but also assumes a bit to much knowledge. On the other hand "to the yet-to-be configured upstream" is way to verbose.)

**Comment by proofit404** [comment]
Seems straightforward to me. But if many people don't understand what push-remote stands for it may results into many annoying issues.

**Comment by tarsius** [comment]
True, but I think that's something I have to go through.

**Comment by tarsius** [comment]
I am considering chancing the binding of "fetch from all remotes" from <kbd>F a</kbd> to <kbd>F F</kbd>. That would restore the same-key-twice binding, but it would do something different from what it used to do before. Instead of fetching from the upstream remote it would fetch from all remotes, but that seems like a pretty harmless change.

**Comment by least-olegs** [comment]
OK, I just found this discussion and here's this:

`P P` was a lot better than the current state of events. If anything, it is easier to make mistakes now, but making mistakes isn't the worst that can happen. Git has idiotic interface. Worst defaults you can imagine and Magit was at least trying to do something useful about the lack of good defaults. `P P` was something you'd expect the versioning system to do. Well, actually, no. You'd expect that committing would do it. No need to mess with the low-level settings every damn time you try to save a chunk of your work. Git added a lot of useless (from the user stand-point) complexity to the process, but this is usually the case with [interfaces designed by engineers](https://pbs.twimg.com/media/BylctRxCEAEv-i2.png).

Now Magit is trying to follow in Git's footsteps? Why? Seriously, sometimes less is more. For anyone who wants to use rarely used features is `:` command.

Ironically, in the effort to make it more "feature reach" you outdid Git itself. Now pushing a new branch to the remote requires **more** typing than you would have done on the command line by using Git proper...

**Comment by Silex** [comment]
@tarsius: just to cheer you up a little, be aware that I find the new push popup way more clearer & less error prone. I don't quite see how someone can claim it is more error prone, but well you were spot on with your tag "deal with the fact that everyone has a different opinion on how this should be done" :) keep up the good work.

**Comment by tarsius** [comment]
You want a version control system with fewer features than Git. That's fine. If you want a commit to be pushed as soon as it is created, then that's your choice and you should probably use a version control system which does that, like subversion for example.

However this is not how Git works and Magit is an interface to Git. Magit is maintained by people who like how Git does certain things, for people who do like Git or at least understand why it does the things it does the way it does them. That doesn't mean that we like everything about Git (or Magit) but it does mean that we have thought about new concepts introduced by Git and other vcs of its generation and have come to the conclusion that we like them. Decoupling committing from pushing, i.e. making version control decentralized, is one of the features we like the most, it makes so many powerful things possible.

Again if you don't want such features, then that is fine. If you think that does features are not just usually useful but actually misguided, then that is fine too. But then Git and Magit just aren't the tools for you. I suggest you instead use Subversion and Emacs' VC package.

It's strange that you want to force your views of how version control ought to work onto a project which is clearly is based on views which conflict with yours.

Also - probably because you don't understand what Magit (and Git) does, and why - you are making some false claims, such as:

> Ironically, in the effort to make it more "feature reach" you outdid Git itself.

No, in the area of pushing, it provides only a subset.

> Now pushing a new branch to the remote requires more typing than you would have done on the command line by using Git proper...

No, <kbd>P p</kbd> and <kbd>P u</kbd> (which are the commands you are supposed to use in 99% of all cases) are both shorter than `git push`. (And yes you first have to configure the upstream resp. the push-remote, but you have to do the same for `git push` to know what it should be doing.)

**Comment by Silex** [comment]
@tarsius: just to cheer you up a little, be aware that I find the new push popup way more clearer & less error prone. I don't quite see how someone can claim it is more error prone, but well you were spot on with your tag "deal with the fact that everyone has a different opinion on how this should be done" :) keep up the good work.

**Comment by tarsius** [comment]
> you were spot on with your tag "deal with the fact that everyone has a different opinion on how this should be done"

Langsam kenne ich meine Pappenheimer :-) But this particular opinion is so far removed from the basic design decisions underlaying Git and by extension Magit, that it simply cannot be taken into account.

**Comment by npostavs** [comment]
> If you want a commit to be pushed as soon as it is created, then that's your choice and you should probably use a version control system which does that, like subversion for example.

If you want a decentralized vc that does this, consider [fossil](http://fossil-scm.org/index.html/doc/trunk/www/index.wiki) (it calls this [autosync mode](http://fossil-scm.org/index.html/doc/trunk/www/concepts.wiki#workflow)).

**Comment by tarsius** [comment]
@least-olegs Since you like `git push` without any arguments so much, you should probably use this:

```lisp
(magit-define-popup-action 'magit-push-popup ?P
  "without specifing what and where\n"
  'magit-push-implicitly)

(defun magit-push-implicitly (args)
  "Push without specifing what to push and where to push it to.

This is a potentially dangerous command.  We recommend that you
don't use it or at least first try with the `--dry-run' argument.

This command simply runs \"git push [ARGS]\" where ARGS are the
arguments specified in the popup buffer.  No other arguments are
used, specifically no arguments are used that control what branch
or branches are to be pushed to what remote and under what name.

The behavior of this command depends on various Git variables:
including, but potentially not limited to, `push.default',
`branch.pushDefault', `branch.<name>.pushRemote'
`branch.<name>.remote' `branch.<name>.merge'."
  (interactive (list (magit-push-arguments)))
  (run-hooks 'magit-credential-hook)
  (magit-run-git-async-no-revert "push" args))
```

**Comment by tarsius** [comment]
@least-olegs Are you **wvxvw** on stackexchange?

**Comment by tarsius** [comment]
Everyone: please give #2444 a try.

**Comment by tarsius** [comment]
> Should same-key-twice bindings be restored?

No, but I will explain how to do that locally.

**Comment by jwiegley** [comment]
Note that `magit-popup` now makes this very easy to add back to Magit:

``` emacs-lisp
(eval-after-load 'magit-remote
    '(progn
       (magit-define-popup-action 'magit-fetch-popup
         ?f 'magit-get-remote #'magit-fetch-from-upstream ?u t)
       (magit-define-popup-action 'magit-pull-popup
         ?F 'magit-get-upstream-branch #'magit-pull-from-upstream ?u t)
       (magit-define-popup-action 'magit-push-popup
         ?P 'magit--push-current-to-upstream-desc
         #'magit-push-current-to-upstream ?u t)))
```

**Comment by jwiegley** [comment]
With these bindings, "doubling" `F` `P` or `f` always refers to upstream.

**Comment by tarsius** [comment]
[Restore pre v2.4 key bindings](https://github.com/magit/magit/wiki/Restore-pre-v2.4-key-bindings) might be of interest too.
